File: solutions/ruby/collatz-conjecture/1/collatz_conjecture.rb
class CollatzConjecture
  
  def self.steps(number)
    raise ArgumentError if number <= 0
    
    result = 0
    
    while number != 1
      if number.even?
        number /= 2
      else
        number = number * 3 + 1
      end
      result += 1
    end
    result
  end
  
end

File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture {

    unsigned int steps(int n) {
        if (n < 1) {
            throw std::domain_error("value must be positive");
        }
        unsigned int steps = 0;

        while (n > 1) {
            n % 2 == 0 ? n /= 2 : n = n * 3 + 1;
            steps += 1;
        }

        return steps;
    }

}
